core/src/main/java/com/github/mizool/core/concurrent/ListenableFutureCollector.java

```diff
@@ -0,0 +1,198 @@
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+
+/**
+ * Collects a {@link java.util.stream.Stream} of {@link ListenableFuture}s of {@link Void} into one.<br>
+ * <br>
+ * Futures are collected as fast as possible without the amount of concurrently running futures exceeding
+ * {@code maximumConcurrentFutures}. The resulting future allows to wait for all futures to complete.<br>
+ * Any {@link java.util.concurrent.ExecutionException} thrown by one of the incoming futures is either thrown during
+ * collection or forwarded to the result future if the consumer was already finished.<br>
+ * <br>
+ * This collector is intended to be used like so:<br>
+ * <pre>{@code
+ * Stream<ListenableFuture<Void>> stream;
+ * stream.collect(ListenableFutureCollector.concurrent(maximumConcurrentFutures)).get();}</pre>
+ * <br>
+ * As the collector can't handle a future's result by design, it only accepts futures without result. If the futures in
+ * the stream <i>do</i> contain results, the intention to discard those results can be documented by using the
+ * {@link ResultVoidingFuture}:<br>
+ * <pre>{@code
+ * Stream<ListenableFuture<MyPojo>> stream;
+ * stream.map(ResultVoidingFuture::new).collect(ListenableFutureCollector.concurrent(maximumConcurrentFutures)).get();}</pre>
+ */
+@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
+public class ListenableFutureCollector implements Collector<ListenableFuture<Void>, Void, ListenableFuture<Void>>
+{
+    public static ListenableFutureCollector concurrent(int maximumConcurrentFutures)
+    {
+        if (maximumConcurrentFutures < 1)
+        {
+            throw new IllegalArgumentException("Maximum concurrent futures must be greater than 0");
+        }
+        return new ListenableFutureCollector(maximumConcurrentFutures);
+    }
+
+    @NoArgsConstructor(access = AccessLevel.PRIVATE)
+    private class Callback implements FutureCallback<Void>
+    {
+        @Override
+        public void onSuccess(Void value)
+        {
+            synchronized (semaphore)
+            {
+                runningFutures--;
+                if (collectionWasFinished())
+                {
+                    setResultValueIfAllFuturesFinished();
+                }
+                semaphore.notifyAll();
+            }
+        }
+
+        @Override
+        public void onFailure(Throwable t)
+        {
+            synchronized (semaphore)
+            {
+                runningFutures--;
+                if (collectionWasFinished())
+                {
+                    result.setException(t);
+                }
+                else
+                {
+                    throwable = t;
+                }
+                semaphore.notifyAll();
+            }
+        }
+    }
+
+    private final int maximumConcurrentFutures;
+    private final Object semaphore = new Object();
+
+    private int runningFutures;
```

Comment:
Use an AtomicInteger instead. Due to the semaphore, the regular int should not pose any problems with its ++ and -- operators, but AtomicIntegers incrementAndGet/decrementAndGet methods will make the intent more clear, I think.